### ISSUE TITLE
[rush] publish tries to upload to default NPM registry when a different registry is listed

### ISSUE THREAD

**Issue by dryzhkov** [microsoft/rushstack#1665]
**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**
'rush publish -p' tries to upload packages to registry.npmjs.org instead of a different registry that's been configured in .npmrc and shrinkwrap.yml

**If the issue is a bug, how can we reproduce it? Please provide detailed steps and include a GitHub branch if applicable. Your issue will get resolved faster if you can make it easy to investigate.**
1. Set a custom (non-default) registry inside .npmrc and shrinkwrap.yml
2. Run `rush publish -p`
3. Observe: the command fails with the following error message 
> npm ERR! 401 Unauthorized - PUT https://registry.npmjs.org/my-app - You must be logged in to publish packages.

**What is the expected behavior?**
Packages should be published to registry specified in step 1

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** rush
* **Tool Version:** 5.17
* **Node Version:** 10.16.0
  * **Is this a LTS version?** yes
  * **Have you tested on a LTS version?** yes
* **OS:** macOS 10.15.1

**Comment by iclanton** [comment]
To publish to a different package feed, use the `--registry <registry>` parameter to `rush publish`. `.npmrc` files allow multiple registries to be specified, so if you're publishing to something other than the default npmjs.org registry, Rush uses an explicit command-line parameter.

**Comment by dryzhkov** [comment]
Confirmed `--registry` parameter solves this issue. However, just to clarify my `.npmrc` config only has 1 (non-default) feed. It seems like rush should respect what's inside `.npmrc`, otherwise, what is the purpose for having this config?

**Comment by octogonz** [comment]
@dryzhkov I believe what you're proposing would be inconsistent with how NPM's `npm publish`  interprets the `.npmrc` file.

Longer term we want to eliminate most of the command-line options for `rush publish` and instead move them into config files (see https://github.com/microsoft/rushstack/issues/904).  But we would use a Rush configuration file (e.g. `verson-policies.json`) since the NPM schemas are somewhat vaguely specified and not validated very well, and they are interpreted differently by different package managers.

**Comment by dryzhkov** [comment]
@octogonz - could you explain why this would be inconsistent with `npm publish`? If `.npmrc` has `registry=https://mycustomregistry.example.org` NPM will publish packages to that registry and not the default (https://registry.npmjs.org). On the other hand, rush seems to try to publish to the default registry regardless of what `.npmrc` has in it and the only way to modify that behavior is to pass `--registry <custom-registry>` as stated above.

**Comment by octogonz** [comment]
> could you explain why this would be inconsistent with `npm publish`? If `.npmrc` has `registry=https://mycustomregistry.example.org` NPM will publish packages to that registry and not the default (https://registry.npmjs.org).

Oh? If that's true, then I'd be okay with changing Rush's behavior to use the default registry specified in `.npmrc`.

**Comment by octogonz** [comment]
Also, note that PR https://github.com/microsoft/rushstack/pull/1602 is introducing a new file `.npmrc-publish` that would be used instead of `.npmrc` when publishing.  So the fix would need to support that.

@dryzhkov would you want to make a PR?

**Comment by dryzhkov** [comment]
I can def give it a try but it might take some time for me to get to it. Maybe #1602 will get merged in the meantime. Is `apps/rush-lib/src/cli/actions/PublishAction.ts` a good starting point for change?

**Comment by octogonz** [comment]
> Is `apps/rush-lib/src/cli/actions/PublishAction.ts` a good starting point for change?

Yup.  And if you need some more direct help figuring out the code, you can chat directly with the Rush developers on [Gitter](https://rushjs.io/pages/help/support/).
